File: src/main/java/by/dev/madhead/jarvis/creator/EmailCreatorFactory.java
package by.dev.madhead.jarvis.creator;

import hudson.FilePath;
import hudson.model.AbstractBuild;
import hudson.model.BuildListener;
import hudson.model.TaskListener;
import org.jenkinsci.plugins.workflow.job.WorkflowRun;

public final class EmailCreatorFactory {

    private EmailCreatorFactory() {
    }

    public static EmailCreator getCreator(AbstractBuild<?, ?> run, BuildListener listener) {
        return new ClassicEmailCreator(run, listener);
    }

    public static EmailCreator getCreator(WorkflowRun run, TaskListener listener, FilePath workspace) {
        return new PipelineEmailCreator(run, listener, workspace);
    }

}

File: src/main/kotlin/by/dev/madhead/jarvis/creator/EmailCreatorFactory.kt
package by.dev.madhead.jarvis.creator

import hudson.FilePath
import hudson.model.AbstractBuild
import hudson.model.BuildListener
import hudson.model.TaskListener
import org.jenkinsci.plugins.workflow.job.WorkflowRun

fun EmailCreator(run: AbstractBuild<*, *>, listener: BuildListener)
        : EmailCreator = ClassicEmailCreator(run, listener)


fun EmailCreator(run: WorkflowRun, listener: TaskListener, workspace: FilePath)
        : EmailCreator = PipelineEmailCreator(run, listener, workspace)
